src/rdkafka_mock_cgrp.c

```diff
@@ -708,3 +723,1014 @@ void rd_kafka_mock_cgrps_connection_closed(rd_kafka_mock_cluster_t *mcluster,
+        else
+                /* Otherwise intersection is empty. */
+                intersection = rd_kafka_topic_partition_list_new(0);
+
+        needs_revocation = member->current_assignment &&
+                           intersection->cnt < member->current_assignment->cnt;
+        if (needs_revocation) {
+                return intersection;
+        }
+
+        rd_kafka_topic_partition_list_destroy(intersection);
+        return NULL;
+}
+
+/**
+ * @brief Calculates if \p member,
+ *        can receive new partitions, given revocation is completed.
+ *        In case new partitions aren't held by other members it
+ *        returns the assignable target assignment and bumps current
+ *        member epoch, otherwise it returns NULL and
+ *        doesn't change current member epoch.
+ *
+ * @param member The group member.
+ *
+ * @return The assignable set of partitions, or NULL in case new partitions
+ *         cannot be assigned yet.
+ *
+ * @remark The returned pointer ownership is transferred to the caller.
+ *
+ * @locks mcluster->lock MUST be held.
+ */
+static rd_kafka_topic_partition_list_t *
+rd_kafka_mock_cgrp_consumer_member_needs_assignment(
+    rd_kafka_mock_cgrp_consumer_member_t *member) {
+        rd_kafka_topic_partition_list_t *returned_assignment =
+            rd_kafka_mock_cgrp_consumer_member_target_assignment_assignable(
+                member);
+
+        if (!rd_kafka_mock_cgrp_consumer_member_next_assignment_can_bump_epoch(
+                member, returned_assignment)) {
+                /* We can't bump the epoch still,
+                 * there are some partitions held by other members.
+                 * We have to return NULL. */
+                rd_kafka_topic_partition_list_destroy(returned_assignment);
+                return NULL;
+        }
+
+        /* No partitions to remove, return
+         * target assignment and reconcile the
+         * epochs */
+        member->current_member_epoch = member->target_member_epoch;
+        return returned_assignment;
+}
+
+/**
+ * @brief Calculates next assignment and member epoch for a \p member,
+ *        given \p current_assignment.
+ *
+ * @param member The group member.
+ * @param current_assignment The assignment sent by the member, or NULL if it
+ *                           didn't change. Must be NULL if *member_epoch is 0.
+ * @param member_epoch Pointer to client reported member epoch. Can be updated.
+ *
+ * @return The new assignment to return to the member.
+ *
+ * @remark The returned pointer ownership is transferred to the caller.
+ *
+ * @locks mcluster->lock MUST be held.
+ */
+rd_kafka_topic_partition_list_t *
+rd_kafka_mock_cgrp_consumer_member_next_assignment(
+    rd_kafka_mock_cgrp_consumer_member_t *member,
+    rd_kafka_topic_partition_list_t *current_assignment,
+    int *member_epoch) {
+        rd_kafka_topic_partition_list_t *returned_assignment = NULL;
```

Comment:
Better to name it `assignments_to_return`.

Reply:
I can change it to `assignment_to_return` as assignment is a set, assignments are a set of sets, for example the assignments for all the group members, while assignment is for a single member.